src/backend/distributed/planner/function_call_delegation.c

```diff
@@ -465,3 +589,202 @@ ShardPlacementForFunctionColocatedWithReferenceTable(CitusTableCacheEntry *cache
+			return;
+		}
+	}
+
+	if (!IsA(partitionValueNode, Const))
+	{
+		/*
+		 * Var node e.g. select fn(col) from table where col=150;
+		 * Param(PARAM_EXEC) node e.g. SELECT fn((SELECT col from test_nested where col=val))
+		 */
+		ereport(ERROR, (errmsg("Non-constant arguments for forcePushdown "
+							   "functions not supported")));
+	}
+	else
+	{
+		Const *partitionValueConst = (Const *) partitionValueNode;
+
+		ereport(DEBUG1, (errmsg("Pushdown argument: %s", pretty_format_node_dump(
+									nodeToString(partitionValueNode)))));
+		EnableInForceDelegatedFuncExecution(partitionValueConst, procedure->colocationId);
+	}
+}
+
+
+/*
+ * Function returns true if the query is simple enough to skip the full executor
+ * It checks only for expressions in the query clauses, and not WHERE and FROM
+ * lists.
+ */
+static bool
+IsQuerySimple(Query *query)
+{
+	if (query->hasAggs ||
+		query->hasWindowFuncs ||
+		query->hasTargetSRFs ||
+		query->hasSubLinks ||
+		query->cteList ||
+		query->groupClause ||
+		query->groupingSets ||
+		query->havingQual ||
+		query->windowClause ||
+		query->distinctClause ||
+		query->sortClause ||
+		query->limitOffset ||
+		query->limitCount ||
+		query->setOperations)
+	{
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * Look for a function in the FROM clause.
+ */
+static FuncExpr *
+FunctionInFromClause(List *fromlist, Query *query)
+{
+	if (list_length(fromlist) != 1)
+	{
+		/* We are looking for a single function */
+		return NULL;
+	}
+
+	RangeTblRef *reference = linitial(fromlist);
+	if (!IsA(reference, RangeTblRef))
+	{
+		/* Skip if there is no RTE */
+		return NULL;
+	}
+
+	RangeTblEntry *rtentry = rt_fetch(reference->rtindex, query->rtable);
+	if (rtentry->rtekind != RTE_FUNCTION)
+	{
+		return NULL;
+	}
+
+	if (list_length(rtentry->functions) != 1)
+	{
+		/* Skip if RTE isn't a single FuncExpr */
+		return NULL;
+	}
+
+	RangeTblFunction *rtfunc = (RangeTblFunction *) linitial(rtentry->functions);
+	if (!IsA(rtfunc->funcexpr, FuncExpr))
+	{
+		/* Skip if RTE isn't a simple FuncExpr */
+		return NULL;
+	}
+
+	return (FuncExpr *) rtfunc->funcexpr;
+}
+
+
+/*
+ * Sets a flag to true indicating that the current node is executing a delegated
+ * function call, using forcePushdown, within a distributed transaction issued
+ * by the coordinator. Also, saves the distribution argument.
+ */
+static void
+EnableInForceDelegatedFuncExecution(Const *distArgument, uint32 colocationId)
+{
+	/*
+	 * The saved distribution argument need to persist through the life
+	 * of the query, both during the planning (where we save) and execution
+	 * (where we compare)
+	 */
+	MemoryContext oldcontext = MemoryContextSwitchTo(TopTransactionContext);
+	ereport(DEBUG1, errmsg("Saving Distribution Argument: %s:%d",
+						   pretty_format_node_dump(nodeToString(distArgument)),
+						   colocationId));
+	AllowedDistributionColumnValue = (AllowedDistributionColumn *) palloc(
+		sizeof(AllowedDistributionColumn));
```

Comment:
You meant, a boolean indicating if the value is set or not, is that right?

Reply:
I meant, we could declare this as a struct rather than a pointer to a struct:
```
AllowedDistributionColumn AllowedDistributionColumnValue;
```
and we set `AllowedDistributionColumnValue.isActive = true` or we just look at `AllowedDistributionColumnValue.distributionColumnValue != NULL` (maybe wrapped in a function)

on the other hand, some allocations are unavoidable given the `copyObject`, so maybe it's not too bad.

one concern I have is that if the transaction has really a lot of forced pushdowns we are leaking memory in the top transaction context. Can we release it earlier?

Reply:
> one concern I have is that if the transaction has really a lot of forced pushdowns we are leaking memory in the top transaction context. Can we release it earlier?

I guess we do release it earlier in the PG_TRY/PG_CATCH, nvm this.

Reply:
Done.